src/category_theory/limits/shapes/cartesian_closed.lean

```diff
@@ -0,0 +1,387 @@
+  map_id' := λ B, pre_id B.unop X,
+  map_comp' := λ P Q R f g, pre_map g.unop f.unop }
+
+lemma pre_post_comm [is_cartesian_closed C] {A B : C} {X Y : Cᵒᵖ} (f : A ⟶ B) (g : X ⟶ Y) :
+  (pre.functor A).map g ≫ post (opposite.unop Y) f = post (opposite.unop X) f ≫ (pre.functor B).map g :=
+begin
+  dsimp [pre],
+  rw [← curry_natural_left, eq_curry_iff, uncurry_natural_right, uncurry_curry, prod_map_map_assoc],
+  simp,
+end
+
+/-- Exponential forms a difunctor. -/
+def exp.difunctor [is_cartesian_closed C] : C ⥤ Cᵒᵖ ⥤ C :=
+{ obj := pre.functor,
+  map := λ A B f, { app := λ X, post X.unop f, naturality' := λ X Y g, pre_post_comm _ _ },
+  map_id' := λ X, by { ext, apply functor.map_id },
+  map_comp' := λ X Y Z f g, by { ext, apply functor.map_comp } }
+
+/-- If an initial object `0` exists in a CCC, then `A ⨯ 0 ≅ 0`. -/
+@[simps]
+def zero_mul [has_initial.{v} C] : A ⨯ ⊥_ C ≅ ⊥_ C :=
+{ hom := limits.prod.snd,
+  inv := default (⊥_ C ⟶ A ⨯ ⊥_ C),
+  hom_inv_id' :=
+  begin
+    have: (limits.prod.snd : A ⨯ ⊥_ C ⟶ ⊥_ C) = uncurry (default _),
+      rw ← curry_eq_iff,
+      apply subsingleton.elim,
+    rw [this, ← uncurry_natural_right, ← eq_curry_iff],
+    apply subsingleton.elim
+  end,
+  }
+
+/-- If an initial object `0` exists in a CCC, then `0 ⨯ A ≅ 0`. -/
+def mul_zero [has_initial.{v} C] : ⊥_ C ⨯ A ≅ ⊥_ C :=
+limits.prod.braiding _ _ ≪≫ zero_mul
+
+/-- If an initial object `0` exists in a CCC then `0^B ≅ 1` for any `B`. -/
+def pow_zero [has_initial.{v} C] [is_cartesian_closed C] : ⊥_C ⟹ B ≅ ⊤_ C :=
+{ hom := default _,
+  inv := curry (mul_zero.hom ≫ default (⊥_ C ⟶ B)),
+  hom_inv_id' :=
+  begin
+    rw [← curry_natural_left, curry_eq_iff, ← cancel_epi mul_zero.inv],
+    { apply subsingleton.elim },
+    { apply_instance },
+    { apply_instance }
+  end }
+
+/--
+If an initial object `0` exists in a CCC then it is a strict initial object,
+i.e. any morphism to `0` is an iso.
+-/
+instance strict_initial [has_initial.{v} C] {f : A ⟶ ⊥_ C} : is_iso f :=
+begin
+  haveI : mono (limits.prod.lift (𝟙 A) f ≫ zero_mul.hom) := mono_comp _ _,
+  rw [zero_mul_hom, prod.lift_snd] at _inst,
+  haveI: split_epi f := ⟨default _, subsingleton.elim _ _⟩,
+  apply is_iso_of_mono_of_split_epi
+end
+
+/-- If an initial object `0` exists in a CCC then every morphism from it is monic. -/
+instance initial_mono (B : C) [has_initial.{v} C] [is_cartesian_closed C] : mono (initial.to B) :=
+⟨λ B g h _, eq_of_inv_eq_inv (subsingleton.elim (inv g) (inv h))⟩
+
+variables {D : Type u₂} [category.{v} D]
+section functor
+
+variables [has_finite_products.{v} D]
+
+/--
+Transport the property of being cartesian closed across an equivalence of categories.
+
+Note we didn't require any coherence between the choice of finite products here, since we transport
+along the `prod_comparison` isomorphism.
+-/
+def cartesian_closed_of_equiv (e : C ≌ D) [h : is_cartesian_closed C] : is_cartesian_closed D :=
+{ cart_closed := λ X,
+  { is_adj :=
+    begin
```

Comment:
This proof will need proper use of `{ ... }`, to we're only working on one goal at a time.

(I was just trying to fix a problem in this proof in my experiment with closed monoidal categories, and this would have made life easier!)